### DISCUSSION TITLE
Equivalence class file

### DISCUSSION THREAD

**Discussion by Ray6283** [COMBINE-lab/salmon#882] [Q&A (bulk specific)] [answered]
From https://salmon.readthedocs.io/en/latest/file_formats.html#quantification-file ,
it says "If salmon was run with the --dumpEq option, then a file called eq_classes.txt will exist in the auxiliary directory"
my question is how do we run with --dumpEq option?
I am following steps described in https://combine-lab.github.io/salmon/getting_started/ 
and could get the results  (the quants file) , but don't know where to put --dumpEq to get the Equivalence class file.
Thanks.

**Comment by rob-p** [comment, marked as answer]
Hi @Ray6283,

  You would simply have to pass `--dumpEq` to the `quant` command.  Then, if your output directory (the argument passed to `-o`) is `<output>`, the file will appear in `<output>/aux_info`.

**Reply by Ray6283** [reply]
Got it. Thanks so much.

**Comment by Ray6283** [comment]
```
Eq class ID    # counts
1 34959 2
7 22632 22633 22634 22635 22636 22637 22638 1928
```

In this case, for equivalence class 1, there is 1 transcript (with 2 fragments), for equivalence class 2, there is  7 transcript (with 1928 fragments to that class)  the last column is number of fragments to that equivalence class.
 I was wondering where these IDs come from? If they are from transcript, do they match with the IDs in transcript file ( in this case I am using the file athal.fa from Salmon page and couldn't match the IDs from equivalence classes to transcripts) .
Thanks

**Reply by Ray6283** [reply]
can we say IDs are just order of transcripts in the file eq_classes.txt?

**Reply by rob-p** [reply]
Yes.  According to the documentation on the equivalence class file format, we have :

```
N (num transcripts)
M (num equiv classes)
tn_1
tn_2
...
tn_N
eq_1_size t_11 t_12 ... count
eq_2_size t_21 t_22 ... count
```

So the ID refers to the order in which the transcript appears in the header of the `eq_classes.txt` file.

**Reply by Ray6283** [reply]
Thanks.

**Reply by Ray6283** [reply]
When run Salmon with --dumpEqWeights or -d , it gives equivalence  classes like this. Is the $P_ij$s same as $w_i^j$s defined in equation 8 in the paper?

N (num transcripts)
M (num equiv classes)
tn_1
tn_2
...
tn_N
eq_1_size t_11 t_12 ... p_11 p_12 ... count
eq_2_size t_21 t_22 ... p_21 p_22 ... count

**Reply by rob-p** [reply]
Those are the weights as defined in [this](https://academic.oup.com/bioinformatics/article/33/14/i142/3953977) paper, which describes our implementation of "range-factorized" equivalence classes.

**Reply by Ray6283** [reply]
Thanks. I couldn't  find how the weights were defined  in [here](https://academic.oup.com/bioinformatics/article/33/14/i142/3953977?login=false) but in [this one](https://www.nature.com/articles/nmeth.4197)  weights are defined . Are they the same?

**Reply by rob-p** [reply]
No, they aren't the same.  In the original salmon paper, within each equivalence class, the weights are simply defined as $\frac{1}{t^i_\ell}$ for transcript $i$.  That is, one over the effective length of the transcript.  However, in the range factorized equivalence class paper, the weights help to *define* the equivalence class.  That is, it is possible to have a equivalence class that consists of the same transcripts but with different weights.

It's not really possible to explain without the paper, but the full definition of the weights is in section 3.2 of the "Improved data-driven likelihood factorizations for transcript abundance estimation" paper I linked.

**Reply by Ray6283** [reply]
Thanks very much @rob-p.

**Reply by Ray6283** [reply]
so sorry, could we say that the probabilities  ( in the equivalence class output file) are 'local' in a sense that are within the equivalence  classes.  Let say we have two transcripts t_i1, and t_i2 (both in in  equivalence class i). Then, 
p_i1 indicates probability that a fragment belongs to t_i1 given  that fragments is from the equivalence class i ( that means it belongs to both t_i1 and t_i2). It's kinda local 'nucleotide fraction'.  I might be wrong  :)
  
However, I couldn't understand when we use equivalence classes,  we assume all the fragments within the same equivalence class get the same probabilities [(equation 6)](https://academic.oup.com/bioinformatics/article/33/14/i142/3953977?login=false) while in the full model we defined bias terms. At the end, which one we use? The full model or the reduced model that uses equivalence classes?

**Comment by Ray6283** [comment]
When we run Salmon with --dumpEqWeights, we get some probabilities in front  of transcripts 
N (num transcripts)
M (num equiv classes)
tn_1
tn_2
...
tn_N
eq_1_size t_11 t_12 ... p_11 p_12 ... count
eq_2_size t_21 t_22 ... p_21 p_22 ... count

Are these $p_{ij}$ same as $w_i^j$ defined in [equation 8](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5600148/pdf/nihms849351.pdf)? Because according to the paper,  $w_i^j$ is average conditional probability of observing a fragment from $l^j$  (equivalence classes)  given $t_i$ over all fragments in this equivalence class.
